### DISCUSSION TITLE
Is it possible to sync the data between three different environments ?

### DISCUSSION THREAD

**Discussion by ErnestManukyan** [Mimetis/Dotmim.Sync#971] [Q&A]
I want to move the data from dev to preprod and then synchronize the preprod with the prod. 

So this is the basic function which I use, and I expect it to synchronize between environments. It works fine for the dev->preprod relation, but the prod db is not changing. When I add changes manually in the preprod then the function is doing synchronization successfully.
The general question is if the framework is able to do synchronization between three environments or no.

```
private async Task SyncBetweenEnvironments(string dev, string preprod, string prod)
        {
            var devProvider = new SqlSyncChangeTrackingProvider(dev);
            var preprodProvider = new SqlSyncChangeTrackingProvider(preprod);
            var prodProvider = new SqlSyncChangeTrackingProvider(prod);

            var options = new SyncOptions { BatchSize = 1000, ConflictResolutionPolicy = ConflictResolutionPolicy.ClientWins };

            var setup = new SyncSetup("dbo.users");
            setup.Tables["dbo.users"].SyncDirection = SyncDirection.UploadOnly;


            var agentDevPreprod = new SyncAgent(devProvider, preprodProvider, options);
            await agentDevPreprod.SynchronizeAsync("devpreprod", setup);

            var agentPreprodProd = new SyncAgent(preprodProvider, prodProvider, options);
            await agentPreprodProd.SynchronizeAsync("preprodprod", setup);
        }
```

Thank you in advance.

**Comment by Mimetis** [comment]
As it's not an issue, I move this to the Q&A discussion section.

**Comment by Mimetis** [comment]
Unfortunately, a database used as **Server** in one scenario and **Client** in another scenario is not currently supported.

It should work, but no guarantee.

I see you are using two scopes, for each scenario / environment, and this is the way to do it, I guess.

Are you able to build a small sample to reproduce your error ? (see this comment on how I create a small repro script : https://github.com/Mimetis/Dotmim.Sync/issues/965#issuecomment-1434647649)

**Reply by ErnestManukyan** [reply]
Sure I can, but I am not getting any errors. The thing is that no any changes are detected by the framework, when relation is from preprod to prod.

**Reply by Mimetis** [reply]
Let's see with a demo script, will be easier to debug.

**Reply by ErnestManukyan** [reply]
Sure, 
Referenced [files](https://github.com/ErnestManukyan/db-examples-dotmim-sync) are the scripts which should create the databases. I am using SQL express. Actually, those are simple databases containing only one [users] table. 
I also provide the function which is running on .net 6

```
       string dev = "Server=localhost\\SQLEXPRESS;Database=dev; TrustServerCertificate=true; Integrated Security=true;";
       string preprod = "Server=localhost\\SQLEXPRESS;Database=preprod; TrustServerCertificate=true; Integrated Security=true;";
       string prod = "Server=localhost\\SQLEXPRESS;Database=prod; TrustServerCertificate=true; Integrated Security=true;";

       await SyncBetweenEnvironments(dev, preprod, prod);

    
        public async Task SyncBetweenEnvironments(string dev, string preprod, string prod)
        {
            var devProvider = new SqlSyncChangeTrackingProvider(dev);
            var preprodProvider = new SqlSyncChangeTrackingProvider(preprod);
            var prodProvider = new SqlSyncChangeTrackingProvider(prod);

            var options = new SyncOptions { BatchSize = 1000, ConflictResolutionPolicy = ConflictResolutionPolicy.ClientWins };

            var setup = new SyncSetup("dbo.users");
            setup.Tables["dbo.users"].SyncDirection = SyncDirection.UploadOnly;


            var agentDevPreprod = new SyncAgent(devProvider, preprodProvider, options);
            await agentDevPreprod.SynchronizeAsync("devpreprod", setup);

            var agentPreprodProd = new SyncAgent(preprodProvider, prodProvider, options);
            await agentPreprodProd.SynchronizeAsync("preprodprod", setup);
        }
```
again thank you in advance

**Reply by Mimetis** [reply]
Thanks.
Can you remove, from your scripts, all the DMS metadata things ?
As it's better to start from scratch when trying to debug

**Reply by Mimetis** [reply]
And in your csharp code, add the steps (insert / Delete / sync / inserts ......) to reproduce the error

**Reply by ErnestManukyan** [reply]
> And in your csharp code, add the steps (insert / Delete / sync / inserts ......) to reproduce the error

you mean add or delete data using c# ?

**Reply by ErnestManukyan** [reply]
[updated files](https://github.com/ErnestManukyan/db-examples-dotmim-sync)
I am adding data manually so I don't have functions which are managing the data. (sorry). I can create, but the case is that I do it manually.
I also turned on the SQL change tracking via this command 
`Alter database dbname SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 14 DAYS, AUTO_CLEANUP = ON)`

**Reply by Mimetis** [reply]
So, please describe the order you are inserting things into your different dbs, and when you are making sync

Unless I have a way to reproduce the error, it will be complicated to reproduce

**Reply by ErnestManukyan** [reply]
I can describe the steps. 
1. Create dev, preprod and prod dbs.
2. Create the same table (users) in the different environments
3. Call the c# function, which does the initial setup. 
4. Manually add a row into the dev database. 
5. Run again the c# function. 

The result is that the new added row in the dev successfully is being synced with the preprod. (The dotmim library moves that new row into the preprod db). But based on the code, it should also do the same thing with the prod db. I am not getting any errors or any messages, but the row is not added into the prod db. When I add a new row into the preprod db manually, then the sync works. I guess the library is not able to detect the changes in preprod db which have been done by the library based on dev changes.

**Reply by Mimetis** [reply]
I've made a quick test, and as I suspected, we can't use a database with two roles (**Server** database, then **Client** database)
